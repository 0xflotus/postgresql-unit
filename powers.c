#include <math.h>
#include <stdio.h>

#define p(n, p, x) printf("#define POWER_%-3d %-22a /* %s %-22a */\n", n, nextafter(x, 0), p, x)

int main ()
{
	printf("/* powers-of-10 constants */\n");
	printf("/* constants here are rounded down so >= comparisons work better */\n\n");


	p(33,  " ", 1e33);
	p(30,  "Q", 1e30);
	p(27,  "R", 1e27);
	p(24,  "Y", 1e24);
	p(21,  "Z", 1e21);
	p(18,  "E", 1e18);
	p(15,  "P", 1e15);
	p(12,  "T", 1e12);
	p(9,   "G", 1e9);
	p(6,   "M", 1e6);
	p(3,   "k", 1e3);
	p(0,   " ", 1e0);
	p(-3,  "m", 1e-3);
	p(-6,  "Âµ", 1e-6);
	p(-9,  "n", 1e-9);
	p(-12, "p", 1e-12);
	p(-15, "f", 1e-15);
	p(-18, "a", 1e-18);
	p(-21, "z", 1e-21);
	p(-24, "y", 1e-24);
	p(-27, "r", 1e-27);
	p(-30, "q", 1e-30);
	p(-33, " ", 1e-33);
}
