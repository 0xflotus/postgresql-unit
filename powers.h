/* powers-of-10 constants */
/* some constants here are rounded down so >= comparisons work better */

#define POWER_27  1e+27                  /*   0x1.9d971e4fe8402p+89  */
#define POWER_24  0x1.a784379d99db3p+79  /* Y 0x1.a784379d99db4p+79  */
#define POWER_21  0x1.b1ae4d6e2ef4fp+69  /* Z 0x1.b1ae4d6e2ef5p+69   */
#define POWER_18  0x1.bc16d674ec7ffp+59  /* E 0x1.bc16d674ec8p+59    */
#define POWER_15  0x1.c6bf52633ffffp+49  /* P 0x1.c6bf52634p+49      */
#define POWER_12  0x1.d1a94a1ffffffp+39  /* T 0x1.d1a94a2p+39        */
#define POWER_9   0x1.dcd64ffffffffp+29  /* G 0x1.dcd65p+29          */
#define POWER_6   0x1.e847fffffffffp+19  /* M 0x1.e848p+19           */
#define POWER_3   0x1.f3fffffffffffp+9   /* k 0x1.f4p+9              */
#define POWER_0   1                      /*   0x1p+0                 */
#define POWER__3  0x1.0624dd2f1a9fbp-10  /* m 0x1.0624dd2f1a9fcp-10  */
#define POWER__6  0x1.0c6f7a0b5ed8cp-20  /* Âµ 0x1.0c6f7a0b5ed8dp-20  */
#define POWER__9  1e-09                  /* n 0x1.12e0be826d695p-30  */
#define POWER__12 0x1.19799812dea10p-40  /* p 0x1.19799812dea11p-40  */
#define POWER__15 1e-15                  /* f 0x1.203af9ee75616p-50  */
#define POWER__18 0x1.2725dd1d243abp-60  /* a 0x1.2725dd1d243acp-60  */
#define POWER__21 0x1.2e3b40a0e9b4ep-70  /* z 0x1.2e3b40a0e9b4fp-70  */
#define POWER__24 0x1.357c299a88ea6p-80  /* y 0x1.357c299a88ea7p-80  */
#define POWER__27 0x1.3ce9a36f23c0fp-90  /*   0x1.3ce9a36f23c1p-90   */
