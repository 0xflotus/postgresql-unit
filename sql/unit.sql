-- test input and output
SELECT '1'::unit;
CREATE TABLE u (u unit);
INSERT INTO u VALUES ('1');
SELECT * FROM u;

-- test constructors
SELECT meter();
SELECT meter(2.0) AS two_meters;
SELECT kilogram();
SELECT kilogram(3.0) AS three_kilogram;
SELECT second();
SELECT second(4.0) AS four_seconds;
SELECT ampere();
SELECT ampere(5.0) AS five_ampere;
SELECT kelvin();
SELECT kelvin(6.0) AS six_kelvin;
SELECT mole();
SELECT mole(7.0) AS seven_mole;
SELECT candela();
SELECT candela(8.0) AS eight_candela;
SELECT byte();
SELECT byte(9.0) AS nine_byte;

-- test operators
SELECT '1'::unit + '2' AS sum;
SELECT '1'::unit - '2' AS difference;
SELECT '2'::unit * '2' AS product;
SELECT '5'::unit / '2' AS fraction;
SELECT '5'::unit / '0' AS division_by_zero;

-- test more operations
SELECT meter() * '2' AS two_meters;
SELECT '3' * meter() AS three_meters;
SELECT meter() * meter() AS square_meter;
SELECT meter() * kilogram() AS meter_kilogram;
SELECT meter() + '1' AS error;
SELECT meter() + kilogram() AS error;

-- test comparisons
WITH
  v(u) AS (VALUES ('1'::unit), ('2'::unit),
	(meter()), (meter() * '2'),
	(kilogram()), (kilogram() * '2')),
  va(a) AS (SELECT * FROM v),
  vb(b) AS (SELECT * FROM v)
SELECT
  a, b, unit_cmp(a, b) AS cmp,
  a < b  AS lt, a <= b AS le,
  a = b  AS eq, a <> b AS ne,
  a >= b AS ge, a > b  AS ge
FROM
  va CROSS JOIN vb;

-- test btree index
INSERT INTO u SELECT meter(generate_series(1,10000)::double precision);
INSERT INTO u SELECT generate_series(1,10000)::text::unit * kilogram();
ANALYZE u;
CREATE INDEX ON u(u);
SELECT * FROM u WHERE u = meter(400);
SELECT * FROM u WHERE u = '300' * kilogram();
EXPLAIN (COSTS OFF) SELECT * FROM u WHERE u = meter(400);
EXPLAIN (COSTS OFF) SELECT * FROM u WHERE u = '300' * kilogram();
