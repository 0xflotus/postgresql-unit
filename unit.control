default_version = '6'
comment = 'unit extension'
relocatable = false
