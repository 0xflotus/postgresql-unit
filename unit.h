#ifndef _UNIT_H
#define _UNIT_H 1

/* indices */
#define UNIT_m		0 /* meter */
#define UNIT_kg		1 /* kilogram */
#define UNIT_s		2 /* second */
#define UNIT_A		3 /* ampere */
#define UNIT_K		4 /* kelvin */
#define UNIT_mol	5 /* mole */
#define UNIT_cd		6 /* candela */
#define UNIT_B		7 /* byte */

#define N_UNITS		8

extern const char *base_units[N_UNITS];

struct derived_unit_t {
	char		*name;
	signed char	 units[N_UNITS];
};

extern const struct derived_unit_t derived_units[];

typedef struct Unit {
	double			value;
	signed char		units[N_UNITS];
} Unit;

int unit_parse (char *s, Unit *unit); /* in unit.y */

#endif /* _UNIT_H */
