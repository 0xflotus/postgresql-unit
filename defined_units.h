#ifndef _DEFINED_UNITS_H
#define _DEFINED_UNITS_H 1

#include <stdio.h> /* NULL */
#include "unit.h"

const char *base_units[N_UNITS] = {
	"m",
	"kg",
	"s",
	"A",
	"K",
	"mol",
	"cd",
	"B",
};

const struct derived_unit_t si_derived_units[] = { /* https://en.wikipedia.org/wiki/International_System_of_Units#Derived_units */
	/* The second is part of this table so we can easily detect its usage for printing it using the hh:mm:ss syntax */
	{ "s",   {  0,  0,  1,  0,  0,  0,  0,  0,} }, /* second          time                                             s */
	{ "Hz",  {  0,  0, -1,  0,  0,  0,  0,  0,} }, /* hertz           frequency                                        s^-1 */
	{ "N",   {  1,  1, -2,  0,  0,  0,  0,  0,} }, /* newton          force, weight                                    kg·m·s^-2 */
	{ "Pa",  { -1,  1, -2,  0,  0,  0,  0,  0,} }, /* pascal          pressure, stress                         N/m^2   kg·m^-1·s^-2 */
	{ "J",   {  2,  1, -2,  0,  0,  0,  0,  0,} }, /* joule           energy, work, heat                       N·m     kg·m^2·s^-2 */
	{ "W",   {  2,  1, -3,  0,  0,  0,  0,  0,} }, /* watt            power, radiant flux                      J/s     kg·m^2·s^-3 */
	{ "C",   {  0,  0,  1,  1,  0,  0,  0,  0,} }, /* coulomb         electric charge                                  s·A */
	{ "V",   {  2,  1, -3, -1,  0,  0,  0,  0,} }, /* volt            voltage                                  W/A     kg·m^2·s^-3·A^-1 */
	{ "F",   { -2, -1,  4,  2,  0,  0,  0,  0,} }, /* farad           electric capacitance                     C/V     kg^-1·m^-2·s^4·A^2 */
	{ "Ω",   {  2,  1, -3, -2,  0,  0,  0,  0,} }, /* ohm             electric resistance, impedance           V/A     kg·m^2·s^-3·A^-2 */
	{ "S",   { -2, -1,  3,  2,  0,  0,  0,  0,} }, /* siemens         electrical conductance                   A/V     kg^-1·m^-2·s^3·A^2 */
	{ "Wb",  {  2,  1, -2, -1,  0,  0,  0,  0,} }, /* weber           magnetic flux                            V·s     kg·m^2·s^-2·A^-1 */
	{ "T",   {  0,  1, -2, -1,  0,  0,  0,  0,} }, /* tesla           magnetic flux density                    Wb/m^2  kg·s^-2·A^-1 */
	{ "H",   {  2,  1, -2, -2,  0,  0,  0,  0,} }, /* henry           inductance                               Wb/A    kg·m^2·s^-2·A^-2 */
	{ "lx",  { -2,  0,  0,  0,  0,  0,  1,  0,} }, /* lux             illuminance                              lm/m^2  m^-2·cd */
	{ "Gy",  {  2,  0, -2,  0,  0,  0,  0,  0,} }, /* gray            absorbed dose (of ionizing radiation)    J/kg    m^2·s^-2 */
	{ "kat", {  0,  0, -1,  0,  0,  1,  0,  0,} }, /* katal           catalytic activity                               mol·s^-1 */
	{ 0 }
};

#endif /* _DEFINED_UNITS_H */
