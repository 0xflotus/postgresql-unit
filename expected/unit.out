-- test input and output
SELECT '1'::unit; -- unit_in
 unit 
------
 1
(1 row)

SELECT unit(1); -- dbl2unit
 unit 
------
 1
(1 row)

CREATE TEMP TABLE u (u unit);
INSERT INTO u VALUES ('1');
SELECT * FROM u;
 u 
---
 1
(1 row)

-- test constructors
SELECT meter();
 meter 
-------
 1 m
(1 row)

SELECT meter(2.0) AS two_meters;
 two_meters 
------------
 2 m
(1 row)

SELECT kilogram();
 kilogram 
----------
 1 kg
(1 row)

SELECT kilogram(3.0) AS three_kilogram;
 three_kilogram 
----------------
 3 kg
(1 row)

SELECT second();
 second 
--------
 1 s
(1 row)

SELECT second(4.0) AS four_seconds;
 four_seconds 
--------------
 4 s
(1 row)

SELECT ampere();
 ampere 
--------
 1 A
(1 row)

SELECT ampere(5.0) AS five_ampere;
 five_ampere 
-------------
 5 A
(1 row)

SELECT kelvin();
 kelvin 
--------
 1 K
(1 row)

SELECT kelvin(6.0) AS six_kelvin;
 six_kelvin 
------------
 6 K
(1 row)

SELECT mole();
 mole  
-------
 1 mol
(1 row)

SELECT mole(7.0) AS seven_mole;
 seven_mole 
------------
 7 mol
(1 row)

SELECT candela();
 candela 
---------
 1 cd
(1 row)

SELECT candela(8.0) AS eight_candela;
 eight_candela 
---------------
 8 cd
(1 row)

SELECT byte();
 byte 
------
 1 B
(1 row)

SELECT byte(9.0) AS nine_byte;
 nine_byte 
-----------
 9 B
(1 row)

-- test dimensionless scales
SELECT unit(0) AS zero, unit(1e-6) AS "1e-6", unit(0.001) AS "0.001", unit(1.0) AS "1", unit(1000.0) AS "1000";
 zero | 1e-6  | 0.001 | 1 | 1000 
------+-------+-------+---+------
 0    | 1e-06 | 0.001 | 1 | 1000
(1 row)

-- test SI prefixes
SELECT meter(0) AS zero, meter(1e-6) AS "µm", meter(0.001) AS mm, meter(1.0) AS m, meter(1000.0) AS km;
 zero |  µm  |  mm  |  m  |  km  
------+------+------+-----+------
 0 m  | 1 µm | 1 mm | 1 m | 1 km
(1 row)

SELECT kilogram(0) AS zero, kilogram(1e-6) AS mg, kilogram(0.001) AS g, kilogram(1.0) AS kg, kilogram(1000.0) AS "Mg";
 zero |  mg  |  g  |  kg  |  Mg  
------+------+-----+------+------
 0 kg | 1 mg | 1 g | 1 kg | 1 Mg
(1 row)

-- test combined units (exactly one unit with exponent 1 in numerator)
SELECT meter(0)/second() AS zero, meter(1e-6)/second() AS "µm", meter(0.001)/second() AS mm, meter(1.0)/second() AS m, meter(1000.0)/second() AS km;
 zero  |   µm   |   mm   |   m   |   km   
-------+--------+--------+-------+--------
 0 m/s | 1 µm/s | 1 mm/s | 1 m/s | 1 km/s
(1 row)

SELECT kilogram(0)/second() AS zero, kilogram(1e-6)/second() AS mg, kilogram(0.001)/second() AS g, kilogram(1.0)/second() AS kg, kilogram(1000.0)/second() AS "Mg";
  zero  |   mg   |   g   |   kg   |   Mg   
--------+--------+-------+--------+--------
 0 kg/s | 1 mg/s | 1 g/s | 1 kg/s | 1 Mg/s
(1 row)

-- test parser
SELECT '1 m'::unit;
 unit 
------
 1 m
(1 row)

SELECT '-1 m/s'::unit;
  unit  
--------
 -1 m/s
(1 row)

SELECT '10 dm^3'::unit, '10l'::unit;
   unit   |   unit   
----------+----------
 0.01 m^3 | 0.01 m^3
(1 row)

SELECT '9.81 kg*m/s^2'::unit, '9.81 kg*m/s*s'::unit, '9.81 kg*m/s/s'::unit;
  unit  |  unit  |  unit  
--------+--------+--------
 9.81 N | 9.81 N | 9.81 N
(1 row)

SELECT '1 foobar'::unit AS error;
ERROR:  unit "foobar" is not known
LINE 1: SELECT '1 foobar'::unit AS error;
               ^
-- test parser arithmetic
SELECT '1|10'::unit, '1|10 m'::unit;
 unit |  unit  
------+--------
 0.1  | 100 mm
(1 row)

SELECT '2 m + 3 m - 4 m'::unit, '6 m - 3 m - 2 m'::unit, '6 m - 3 m + 1 m'::unit;
 unit | unit | unit 
------+------+------
 1 m  | 1 m  | 4 m
(1 row)

SELECT '2 m * 3 m'::unit, '2 m / 3 m'::unit, '10 m / 5 m / 2 m'::unit;
 unit  |       unit        |  unit  
-------+-------------------+--------
 6 m^2 | 0.666666666666667 | 1 m^-1
(1 row)

SELECT '2 m^2 + 3 m * 4 m'::unit, '2 m * 3 m / 4 m * 5 m'::unit;
  unit  | unit 
--------+------
 14 m^2 | 0.3
(1 row)

SELECT '2 m * (1 m + 3 m)'::unit;
 unit  
-------
 8 m^2
(1 row)

SELECT '- m'::unit, '/ m'::unit;
 unit |  unit  
------+--------
 -1 m | 1 m^-1
(1 row)

SELECT '4+5'::unit, '4-5'::unit, '4(-5)'::unit, '4*5'::unit, '4/5'::unit;
 unit | unit | unit | unit | unit 
------+------+------+------+------
 9    | -1   | -20  | 20   | 0.8
(1 row)

-- problematic cases
SELECT '1 cd'::unit AS candela; -- candela vs centiday
 candela 
---------
 1 cd
(1 row)

SELECT '1 Pa'::unit AS pascal; -- pascal vs petayear
 pascal 
--------
 1 Pa
(1 row)

SELECT '1 yg'::unit AS yoctogram, '1.00001 yg'::unit AS actual_yoctogram; -- "yg" had rounding problems in the past
 yoctogram | actual_yoctogram 
-----------+------------------
 1 yg      | 1.00001 yg
(1 row)

SELECT '1 min'::unit AS minute; -- minute vs milliinch
   minute   
------------
 00:01:00 s
(1 row)

SELECT '1 ft'::unit AS foot; -- foot vs femtotonne
   foot   
----------
 304.8 mm
(1 row)

SELECT '1 yd'::unit AS yard; -- yard vs yoctoday
   yard   
----------
 914.4 mm
(1 row)

-- units whose definition changed between v2 and 3
SELECT 'a'::unit AS are;
   are   
---------
 100 m^2
(1 row)

SELECT 'da'::unit AS day;
 day 
-----
 1 d
(1 row)

SELECT 'rad'::unit AS radiation;
 radiation 
-----------
 10 mGy
(1 row)

-- check which prefix/unit combinations produce other dimensions (parser conflicts)
SELECT
  p, u, (p||u)::unit
FROM
  prefixes CROSS JOIN units
WHERE
  u <> 'kg' AND
  dimension(u::unit) != dimension((p||u)::unit) AND
  p||u NOT IN ('dat') -- skip ambiguous unit
ORDER BY lower(p) COLLATE "C", lower(u) COLLATE "C";
 p |  u  |             unit              
---+-----+-------------------------------
 a | t   | 98.0665 kPa
 c | d   | 1 cd
 c | h   | 20.1168 m
 d | a   | 1 d
 f | t   | 304.8 mm
 G | s   | 100 µT
 h | bar | 1.05457180005237e-34 m^2*kg/s
 h | d   | 0.238480942392 m^3
 k | in  | 600 g
 m | h   | 1 mH
 m | in  | 00:01:00 s
 n | t   | 1 N
 P | a   | 1 Pa
 p | a   | 1 Pa
 p | h   | 10 klx
 p | in  | 0.020457405 m^3
 p | t   | 0.000473176473 m^3
 y | d   | 914.4 mm
(18 rows)

